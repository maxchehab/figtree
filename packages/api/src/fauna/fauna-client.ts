import { Client, query as q, ExprVal } from 'faunadb';

import { FaunaEntity } from './interfaces/fauna-entity.interface';
import { LoginRequest } from '../common/interfaces/login-request.interface';
import { Project } from '../common/interfaces/project.interface';
import { User } from '../common/interfaces/user.interface';

export class FaunaClient {
  private readonly client = new Client({
    secret: process.env.FAUNA_SECRET as string,
  });

  async userByLoginRequest(code: string): Promise<FaunaEntity<User> | null> {
    try {
      return await this.client.query(
        q.Get(q.Match(q.Index('users_by_login_request_code'), code)),
      );
    } catch (_error) {
      return null;
    }
  }

  async updateUser(ref: ExprVal, data: Partial<User>) {
    return this.client.query(q.Update(ref, { data }));
  }

  async userByToken(token: string): Promise<FaunaEntity<User> | null> {
    try {
      return await this.client.query(
        q.Get(q.Match(q.Index('users_by_token'), token)),
      );
    } catch (_error) {
      return null;
    }
  }

  async userByID(id: string): Promise<FaunaEntity<User> | null> {
    try {
      return await this.client.query(
        q.Get(q.Match(q.Index('users_by_id'), id)),
      );
    } catch (_error) {
      return null;
    }
  }

  async projectsByUser(
    user: FaunaEntity<User>,
  ): Promise<Array<FaunaEntity<Project>>> {
    return this.client.query(
      q.Map(
        q.Paginate(q.Match(q.Index('projects_by_user_id'), user.data.id)),
        q.Lambda('X', q.Get(q.Var('X'))),
      ),
    );
  }

  async createUser(data: Partial<User>): Promise<FaunaEntity<User>> {
    return this.client.query(
      q.Create(q.Collection('users'), {
        data,
      }),
    );
  }

  async createLoginRequest(
    code: string,
    createdAt: Date,
  ): Promise<FaunaEntity<LoginRequest>> {
    return this.client.query(
      q.Create(q.Collection('login_requests'), {
        data: { code, createdAt: createdAt.toISOString() },
      }),
    );
  }

  async loginRequestByCode(
    code: string,
  ): Promise<FaunaEntity<LoginRequest> | null> {
    try {
      return await this.client.query(
        q.Get(q.Match(q.Index('login_requests_by_code'), code)),
      );
    } catch (_error) {
      return null;
    }
  }
}
