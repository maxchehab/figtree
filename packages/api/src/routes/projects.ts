import { AuthStatus } from '../common/interfaces/auth-status.enum';
import { FaunaClient } from '../fauna/fauna-client';
import { HttpException } from '../common/exceptions/http.exception';
import getBearerToken from '../common/util/get-bearer-token.util';
import lambda from '../common/util/lambda.util';

const faunaClient = new FaunaClient();

export default lambda(async (req, res) => {
  const token = getBearerToken(req);
  const user = await faunaClient.userByToken(token);

  if (!user) {
    throw new HttpException(400, 'Invalid bearer token.');
  }

  const projects = await faunaClient.projectsByUser(user);

  return res.status(200).json({ projects: projects.map(({ data }) => data) });
});
