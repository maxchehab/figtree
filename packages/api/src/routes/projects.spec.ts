import axios from 'axios';
import http from 'http';
import listen from 'test-listen';

import { AuthStatus } from '../common/interfaces/auth-status.enum';
import { FaunaClient } from '../fauna/fauna-client';
import closeServer from '../test/close-server';
import createTestLambda from '../test/create-test-lambda';
import projects from './projects';

describe('projects', () => {
  let server: http.Server;
  let url: string;

  beforeAll(async () => {
    server = createTestLambda(projects);
    url = await listen(server);
  });

  afterAll(async () => {
    await closeServer(server);
  });

  describe('with no user associated to the token', () => {
    beforeEach(() => {
      jest
        .spyOn(FaunaClient.prototype, 'userByToken')
        .mockImplementationOnce(async () => null);
    });

    it('returns 400, Invalid bearer token', async () => {
      const { status, data } = await axios.get(url, {
        validateStatus: () => true,
        headers: { authorization: 'Bearer invalid_token' },
      });

      expect(status).toEqual(400);
      expect(data).toEqual({
        message: 'Invalid bearer token.',
      });
    });
  });

  describe('with two projects associated with the user', () => {
    beforeEach(() => {
      jest
        .spyOn(FaunaClient.prototype, 'userByToken')
        .mockImplementationOnce(async () => ({
          ref: '@ref:123',
          data: {
            email: 'example@gmail.com',
            auth_status: AuthStatus.LoggedOut,
            id: 'usr_123',
            login_request_code: 'code_123',
            token: 'token_123',
          },
        }));

      jest
        .spyOn(FaunaClient.prototype, 'projectsByUser')
        .mockImplementationOnce(async () => [
          {
            ref: '@ref:123',
            data: {
              id: 'code_123',
              name: 'token_123',
              user_id: 'usr_123',
            },
          },
          {
            ref: '@ref:123',
            data: {
              id: 'code_123',
              name: 'token_123',
              user_id: 'usr_123',
            },
          },
        ]);
    });

    it('returns two projects', async () => {
      const { status, data } = await axios.get(url, {
        validateStatus: () => true,
        headers: { authorization: 'Bearer token_123' },
      });

      expect(status).toEqual(200);
      expect(data.projects.length).toEqual(2);
      expect(data.projects).toMatchSnapshot();
    });
  });

  describe('with two projects associated with the user', () => {
    beforeEach(() => {
      jest
        .spyOn(FaunaClient.prototype, 'userByToken')
        .mockImplementationOnce(async () => ({
          ref: '@ref:123',
          data: {
            email: 'example@gmail.com',
            auth_status: AuthStatus.LoggedOut,
            id: 'usr_123',
            login_request_code: 'code_123',
            token: 'token_123',
          },
        }));

      jest
        .spyOn(FaunaClient.prototype, 'projectsByUser')
        .mockImplementationOnce(async () => []);
    });

    it('returns empty projects array', async () => {
      const { status, data } = await axios.get(url, {
        validateStatus: () => true,
        headers: { authorization: 'Bearer token_123' },
      });

      expect(status).toEqual(200);
      expect(data).toEqual({ projects: [] });
    });
  });
});
