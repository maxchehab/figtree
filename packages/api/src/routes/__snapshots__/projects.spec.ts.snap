// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projects with two projects associated with the user returns two projects 1`] = `
Array [
  Object {
    "id": "code_123",
    "name": "token_123",
    "user_id": "usr_123",
  },
  Object {
    "id": "code_123",
    "name": "token_123",
    "user_id": "usr_123",
  },
]
`;
